### PR TITLE
Disabled stream mgmt crashes

### PR DESCRIPTION
**Pull request by NelsonVides**
If `mod_stream_management` is disabled, the session shouldn't crash if it requests so. Rather, the server should simply respond with a`service-unavailable` and continue.

Currently, as the code for stream management is mostly hardcoded in ejabberd_c2s, if the client sends the `<<"enable">>` stanza, the server goes nuts:
https://github.com/esl/MongooseIM/blob/1fa0898d6901549e5df5a646dcbd95a7c05b9e7d/src/ejabberd_c2s.erl#L2737-L2749

If the user requests stream management without resumption, the server answers with a nonsensical `<<"enabled">>` and continues as if nothing happened, until it crashes, and if the client requests so but _with_ resumption, then his c2s process fails fast.

TODO:
- [x] FIX IT! This is only a regression test for now 😛

### CODE REVIEW

**Comment by mongoose-im** [comment]
[7288.1](https://travis-ci.org/esl/MongooseIM/jobs/592563543) / Erlang 22.0 / small_tests / 9cdb5e228b457d8ae662ce34dc0c4f1df4bcc067
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2482/7288/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2482/7288/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-10-02_13.48.24/lib.mongooseim.logs/run.2019-10-02_13.48.25/suite.log.html)

**Comment by mongoose-im** [comment]
[7305.1](https://travis-ci.org/esl/MongooseIM/jobs/597548394) / Erlang 22.0 / small_tests / d7ce8c9413dc7b90c814d1e15cfef4c811a5b328
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2482/7305/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2482/7305/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-10-14_09.52.26/lib.mongooseim.logs/run.2019-10-14_09.52.27/suite.log.html)

**Comment by mongoose-im** [comment]
[7307.1](https://travis-ci.org/esl/MongooseIM/jobs/597581394) / Erlang 22.0 / small_tests / cc0a003406da0974b34def63177af11682a7f946
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2482/7307/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2482/7307/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-10-14_12.10.39/lib.mongooseim.logs/run.2019-10-14_12.10.39/suite.log.html)

File: big_tests/tests/sm_SUITE.erl [hunk 1]
```diff
@@ -159,6 +174,11 @@ register_some_smid_h(Config) ->
     TestSmids = lists:map(fun register_smid/1, lists:seq(1, 3)),
     [{smid_test, TestSmids} | Config].
 
+init_per_testcase(no_crash_if_stream_mgmt_disabled = CN, Config) ->
+    Config2 = dynamic_modules:save_modules(domain(), Config),
```
Comment (fenek) [inline big_tests/tests/sm_SUITE.erl]: Isn't it a duplicate of init for a group?

File: big_tests/tests/sm_SUITE.erl [hunk 2]
```diff
@@ -1075,6 +1098,26 @@ messages_are_properly_flushed_during_resumption(Config) ->
         escalus:assert(is_chat_message, [MsgBody], RecvMsg)
       end).
 
+no_crash_if_stream_mgmt_disabled_but_client_requests_stream_mgmt(Config) ->
+    AliceSpec = escalus_fresh:create_fresh_user(Config, alice),
+    Steps = connection_steps_to_session(),
+    {ok, Alice, _Features} = escalus_connection:start(AliceSpec, Steps),
+    %% Should not crash anything!
+    escalus_connection:send(Alice, escalus_stanza:enable_sm()),
+    % escalus_connection:send(Alice, escalus_stanza:enable_sm([resume])),
```
Comment (fenek) [inline big_tests/tests/sm_SUITE.erl]: May be deleted.

File: src/ejabberd_c2s.erl [hunk 3]
```diff
@@ -204,6 +204,10 @@ init([{SockMod, Socket}, Opts]) ->
     TLSOpts = verify_opts(Verify) ++ TLSOpts1,
     [ssl_crl_cache:insert({file, CRL}) || CRL <- proplists:get_value(crlfiles, Opts, [])],
     IP = peerip(SockMod, Socket),
+    StreamMgmtConfig = case gen_mod:is_loaded(?MYNAME, mod_stream_management) of
```
Comment (fenek) [inline src/ejabberd_c2s.erl]: This should be done when we actually know the proper server. We can afford to wait. :)

File: big_tests/tests/sm_SUITE.erl [hunk 4]
```diff
@@ -130,9 +138,16 @@ init_per_group(parallel_manual_ack_freq_1, Config) ->
     escalus_users:update_userspec(Config, alice, manual_ack, true);
 init_per_group(stale_h, Config) ->
     escalus_users:update_userspec(Config, alice, manual_ack, true);
+init_per_group(stream_mgmt_disabled, Config) ->
+    Config2 = dynamic_modules:save_modules(domain(), Config),
+    rpc(mim(), gen_mod, stop_module, [domain(), ?MOD_SM]),
```
[the remark this answers is not in the document]
Comment (NelsonVides) [reply big_tests/tests/sm_SUITE.erl]: Damn, I'm blind. Done! 😅

File: big_tests/tests/sm_SUITE.erl (hunk 1 above)
Comment (NelsonVides) [reply big_tests/tests/sm_SUITE.erl]: Indeed, there's no such testcase even. This was my initial test, until I realised there are two different conditions to test and then made it into a group. Removing this trash, good catch!

**Comment by mongoose-im** [comment]
[7310.1](https://travis-ci.org/esl/MongooseIM/jobs/598186926) / Erlang 22.0 / small_tests / e38c714f1944d0d792a59479d5b3aa2718e27c87
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2482/7310/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2482/7310/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-10-15_14.44.15/lib.mongooseim.logs/run.2019-10-15_14.44.16/suite.log.html)

File: src/ejabberd_c2s.erl [hunk 5]
```diff
@@ -272,8 +267,13 @@ wait_for_stream(closed, StateData) ->
 
 handle_stream_start({xmlstreamstart, _Name, Attrs}, #state{} = S0) ->
     Server = jid:nameprep(xml:get_attr_s(<<"to">>, Attrs)),
+    ?ERROR_MSG("Server is ~p; MYNAME is ~p~n", [Server, ?MYNAME]),
```
Comment (fenek) [inline src/ejabberd_c2s.erl]: WTF? Error level?! :D

Comment (NelsonVides) [reply src/ejabberd_c2s.erl]: FUCK, that's me trying to see the value. and forgetting to remove it when committing xD

**Comment by mongoose-im** [comment]
[7311.1](https://travis-ci.org/esl/MongooseIM/jobs/598534097) / Erlang 22.0 / small_tests / 372fae8d4609b61cbc0ea087bf17e975860479f9
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2482/7311/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2482/7311/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-10-16_08.19.13/lib.mongooseim.logs/run.2019-10-16_08.19.14/suite.log.html)

**Pull request merged**